Port gpfdist sink

jvalkeal opened the pull request:
- Port code and tests as it from xd module into
  a boot app.
- Integration tests are disabled by default, enable with
  `-DskipITs=false`.
- Fixes #74 

From psql:

```
gpadmin=# create table ticktock (date text, time text);
```

Simple test assuming gpdb master host is `mdw`

```
# java -jar time-source/target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar --server.port=8081 --spring.cloud.stream.bindings.output=xxx
# java -jar gpfdist-sink/target/gpfdist-sink-1.0.0.BUILD-SNAPSHOT-exec.jar --server.port=8082 --spring.cloud.stream.bindings.input=xxx --dbHost=mdw --table=ticktock --batchTime=1 --batchPeriod=1 --flushCount=2 --flushTime=2 --columnDelimiter=" "
```

Check results:

```
gpadmin=# select count(*) from ticktock;
 count 
-------
   152
(1 row)
```

jvalkeal committed:
Port gpfdist sink

- Port code and tests as it from xd module into
  a boot app.
- Integration tests are disabled by default, enable with
  `-DskipITs=false`.
- Fixes #74

artembilan commented on gpfdist-sink/src/main/java/org/springframework/cloud/stream/module/gpfdist/sink/AbstractGpfdistMessageHandler.java:
```diff
@@ -0,0 +1,175 @@
+ *
+ * @author Janne Valkealahti
+ */
+public abstract class AbstractGpfdistMessageHandler extends AbstractMessageHandler implements SmartLifecycle {
```
No one `MessageHandler` can be `SmartLifecycle`. That is just `Lifecycle` which is propagated from its `Endpoint`:

```
@Override
protected void doStart() {
    this.logComponentSubscriptionEvent(true);
    this.inputChannel.subscribe(this.handler);
    if (this.handler instanceof Lifecycle) {
        ((Lifecycle) this.handler).start();
    }
}
```

Otherwise we may end up with the double `start()/stop()` calls.

jvalkeal replied:
start() is protected by a flag, stop() would not fail.

artembilan replied:
Well, I don't remember already "why?", but that is how we reworked all our `MessageHandler`s and `MessageSource`s.
Maybe the main point do not start automatically target object before the endpoint around it...
CC @garyrussell

trisberg commented:
Trying to run a simple ticktock test and get this:

```
2016-01-14 13:05:22.510 ERROR 84148 --- [TaskScheduler-1] o.s.c.s.m.g.sink.GpfdistMessageHandler   : Error in load

org.springframework.jdbc.BadSqlGrammarException: StatementCallback; bad SQL grammar [DROP EXTERNAL TABLE ticktock_ext_bc705973_b265_4542_aa38_d18b923c96ad]; nested exception is org.postgresql.util.PSQLException: ERROR: table "ticktock_ext_bc705973_b265_4542_aa38_d18b923c96ad" does not exist
	at org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator.doTranslate(SQLErrorCodeSQLExceptionTranslator.java:231) ~[spring-jdbc-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:73) ~[spring-jdbc-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:415) ~[spring-jdbc-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:439) ~[spring-jdbc-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
	at org.springframework.cloud.stream.module.gpfdist.sink.support.JdbcCommands.clean(JdbcCommands.java:148) ~[gpfdist-sink-1.0.0.BUILD-SNAPSHOT-exec.jar!/:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.stream.module.gpfdist.sink.support.JdbcCommands.execute(JdbcCommands.java:103) ~[gpfdist-sink-1.0.0.BUILD-SNAPSHOT-exec.jar!/:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.stream.module.gpfdist.sink.support.DefaultLoadService.load(DefaultLoadService.java:57) ~[gpfdist-sink-1.0.0.BUILD-SNAPSHOT-exec.jar!/:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.stream.module.gpfdist.sink.support.DefaultGreenplumLoad.load(DefaultGreenplumLoad.java:45) ~[gpfdist-sink-1.0.0.BUILD-SNAPSHOT-exec.jar!/:1.0.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.stream.module.gpfdist.sink.GpfdistMessageHandler$1.run(GpfdistMessageHandler.java:141) ~[gpfdist-sink-1.0.0.BUILD-SNAPSHOT-exec.jar!/:1.0.0.BUILD-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) [na:1.7.0_79]
	at java.util.concurrent.FutureTask.run(FutureTask.java:262) [na:1.7.0_79]
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54) [spring-context-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) [na:1.7.0_79]
	at java.util.concurrent.FutureTask.run(FutureTask.java:262) [na:1.7.0_79]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178) [na:1.7.0_79]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292) [na:1.7.0_79]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_79]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_79]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_79]
Caused by: org.postgresql.util.PSQLException: ERROR: table "ticktock_ext_bc705973_b265_4542_aa38_d18b923c96ad" does not exist
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2182) ~[postgresql-9.4-1206-jdbc41.jar!/:9.4]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1911) ~[postgresql-9.4-1206-jdbc41.jar!/:9.4]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:173) ~[postgresql-9.4-1206-jdbc41.jar!/:9.4]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:645) ~[postgresql-9.4-1206-jdbc41.jar!/:9.4]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:481) ~[postgresql-9.4-1206-jdbc41.jar!/:9.4]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:473) ~[postgresql-9.4-1206-jdbc41.jar!/:9.4]
	at org.apache.commons.dbcp.DelegatingStatement.execute(DelegatingStatement.java:264) ~[commons-dbcp-1.4.jar!/:1.4]
	at org.apache.commons.dbcp.DelegatingStatement.execute(DelegatingStatement.java:264) ~[commons-dbcp-1.4.jar!/:1.4]
	at org.springframework.jdbc.core.JdbcTemplate$1ExecuteStatementCallback.doInStatement(JdbcTemplate.java:431) ~[spring-jdbc-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:404) ~[spring-jdbc-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
	... 16 common frames omitted

```

trisberg commented:
Also, can't seem to stop the module, have to kill -9 it

jvalkeal commented:
That error on its own doesn't make that much sense. If it can't drop table at the end, there has to be other error telling table could not be created. DB permissions issues?

trisberg commented:
had to add --dbName parameter, we need to improve the error logging so we see more than the last error

trisberg commented on gpfdist-sink/src/main/java/org/springframework/cloud/stream/module/gpfdist/sink/support/JdbcCommands.java:
```diff
@@ -0,0 +1,199 @@
+			jdbcTemplate.execute(prepareSql);
+		}
+		catch (DataAccessException e) {
+			lastException = e;
```
we should log this error

trisberg commented on gpfdist-sink/src/main/java/org/springframework/cloud/stream/module/gpfdist/sink/support/JdbcCommands.java:
```diff
@@ -0,0 +1,199 @@
+			jdbcTemplate.execute(runSql);
+		}
+		catch (DataAccessException e) {
+			lastException = e;
```
we should log this error

jvalkeal committed:
Add error logging for JdbcCommands

- Instead of keeping last error during prepare,
  run and clean, log all exceptions.
- Relates to #74

trisberg commented:
Merged via commit 1898867c155ffc58edf4dc864a1364e85b4ff196 and commit aff61c95dcec87d42681d4a1bd6cf4fc31314f62
